### PR TITLE
feat: add setter and getter apis to specify udp port range for webrtc

### PR DESCRIPTION
**Pull request by wanted002**
#### Description of Change
Add 'getWebRTCUDPPortRange' and 'setWebRTCUDPPortRange'  apis to specify udp port range for webrtc, which is part of a feature described in [issue#9042.](https://github.com/electron/electron/issues/9042) Command line configuration for the WebRtcUdpPortRange policy will add later.

#### Checklist

- [x] `npm test` passes
- [x] relevant documentation, tutorials, templates and examples are changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/main/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense]

#### Release Notes

Notes: Added 'getWebRTCUDPPortRange' and 'setWebRTCUDPPortRange'  apis to specify udp port range for webrtc.

### CODE REVIEW

File: shell/browser/api/electron_api_web_contents.cc [hunk 1]
```diff
@@ -2536,6 +2577,28 @@ void WebContents::SetWebRTCIPHandlingPolicy(
...
+const std::string WebContents::GetWebRTCUDPPortRangePolicy(void) const {
+  return ToPortRange(
+      web_contents()->GetMutableRendererPrefs()->webrtc_udp_min_port,
+      web_contents()->GetMutableRendererPrefs()->webrtc_udp_max_port);
+}
+
+void WebContents::SetWebRTCUDPPortRangePolicy(
+    const std::string& webrtc_udp_port_range_policy) {
+  uint16_t port_min = 0;
+  uint16_t port_max = 0;
```
Comment (codebytere) [inline shell/browser/api/electron_api_web_contents.cc:2589]: Given we'll never use these variables if the conditional on line 2591 evaluates to false, they should be moved below the conditional.

File: spec/api-web-contents-spec.ts [hunk 2]
```diff
@@ -1288,6 +1288,22 @@ describe('webContents module', () => {
     });
   });
 
+  describe('webrtc udp port range policy api', () => {
+    afterEach(closeAllWindows);
```
Comment (codebytere) [inline spec/api-web-contents-spec.ts:1292]: In conjunction with my comments about silent failure above, we should also be testing invalid port ranges to ensure that they throw correctly and descriptively.

**Review by codebytere — CHANGES_REQUESTED** [review]
Thanks for working on this! I've left some implementation and API design comments but i'd be happy to see this move forward.

File: shell/browser/api/electron_api_web_contents.cc [hunk 3]
```diff
@@ -718,6 +718,47 @@ content::RenderFrameHost* GetRenderFrameHost(
...
 
+std::string ToPortRange(uint16_t min_port, uint16_t max_port) {
+  if (min_port == 0 || min_port > max_port || max_port > UINT16_MAX) {
+    return std::string();
+  }
+
+  return base::NumberToString(min_port) + "-" + base::NumberToString(max_port);
+}
+
+void ParsePortRange(const std::string& range,
```
[the remark this answers is not in the document]
Comment (wanted002) [reply shell/browser/api/electron_api_web_contents.cc:729]: Ok, this is reasonable, and thanks very much for your patiently guide :)

File: shell/browser/api/electron_api_web_contents.cc [hunk 4]
```diff
@@ -2536,6 +2577,28 @@ void WebContents::SetWebRTCIPHandlingPolicy(
   web_contents()->SyncRendererPrefs();
 }
 
+const std::string WebContents::GetWebRTCUDPPortRangePolicy(void) const {
+  return ToPortRange(
+      web_contents()->GetMutableRendererPrefs()->webrtc_udp_min_port,
+      web_contents()->GetMutableRendererPrefs()->webrtc_udp_max_port);
```
[the remark this answers is not in the document]
Comment (wanted002) [reply shell/browser/api/electron_api_web_contents.cc:2583]: I got be lazy, sure ,I'll fix this

File: shell/browser/api/electron_api_web_contents.cc [hunk 5]
```diff
@@ -2536,6 +2577,28 @@ void WebContents::SetWebRTCIPHandlingPolicy(
...
+
+void WebContents::SetWebRTCUDPPortRangePolicy(
+    const std::string& webrtc_udp_port_range_policy) {
+  uint16_t port_min = 0;
+  uint16_t port_max = 0;
+
+  if (GetWebRTCUDPPortRangePolicy() == webrtc_udp_port_range_policy)
+    return;
+
+  ParsePortRange(webrtc_udp_port_range_policy, &port_min, &port_max);
```
[the remark this answers is not in the document]
Comment (wanted002) [reply shell/browser/api/electron_api_web_contents.cc:2594]: ok, got it. Many thanks for your review and guide to a newbie like me~

**Comment by wanted002** [comment]
Hello~~ Could anyone retrigger the linux build,please? It seems the compiler proxy didn't work at that time.

File: docs/api/web-contents.md [hunk 6]
```diff
@@ -2064,6 +2064,21 @@ Setting the WebRTC IP handling policy allows you to control which IPs are
 exposed via WebRTC. See [BrowserLeaks](https://browserleaks.com/webrtc) for
 more details.
 
+#### `contents.getWebRTCUDPPortRangePolicy()`
+
+Returns `Object`:
+
+* `min` Integer - The minimum port number of WebRTC UDP Port Range Policy.
+* `max` Integer - The maximum port number of WebRTC UDP Port Range Policy.
```
Comment (itsananderson) [inline docs/api/web-contents.md:2072]: Poking around in Chromium internals, it sounds like these would default to `{ min: 0, max: 0 }` which would apply no restriction on port range. Should we include a note (and maybe a unit test) about the default value/behavior of `setWebRTCUDPPortRangePolicy` hasn't been called?

File: docs/api/web-contents.md [hunk 7]
```diff
@@ -2064,6 +2064,21 @@ Setting the WebRTC IP handling policy allows you to control which IPs are
...
+
+#### `contents.setWebRTCUDPPortRangePolicy(udpPortRange)`
+
+* `udpPortRange` Object
+  * `min` Integer - The minimum port number of WebRTC UDP Port Range Policy.
+  * `max` Integer - The maximum port number of WebRTC UDP Port Range Policy.
+
+Setting the WebRTC UDP Port Range policy allows you to restrict the udp port range assigned to apps via WebRTC.
```
Comment (itsananderson) [inline docs/api/web-contents.md:2080]: If someone sets a policy and then wants to remove it, would they pass `{ min: 0, max: 0 }`? Is it worth including that in the docs?

File: docs/api/web-contents.md (hunk 6 above)
Comment (wanted002) [reply docs/api/web-contents.md:2072]: Yeah, a note for the default behavior is reasonable

File: docs/api/web-contents.md (hunk 7 above)
Comment (wanted002) [reply docs/api/web-contents.md]: Sure, this is a reasonable scene, thanks for your review.

**Comment by wanted002** [comment]
Sorry for the mistaking merge operation. Fixed it. Please review for me ,thanks

File: docs/api/web-contents.md [hunk 8]
```diff
@@ -2064,6 +2064,23 @@ Setting the WebRTC IP handling policy allows you to control which IPs are
 exposed via WebRTC. See [BrowserLeaks](https://browserleaks.com/webrtc) for
 more details.
 
+#### `contents.getWebRTCUDPPortRangePolicy()`
```
Comment (jkleinsc) [inline docs/api/web-contents.md:2067]:
```suggestion
#### `contents.getWebRTCUDPPortRange()`
```

File: docs/api/web-contents.md [hunk 9]
```diff
@@ -2064,6 +2064,23 @@ Setting the WebRTC IP handling policy allows you to control which IPs are
...
+
+Returns `Object`:
+
+* `min` Integer - The minimum port number of WebRTC UDP Port Range Policy.
+* `max` Integer - The maximum port number of WebRTC UDP Port Range Policy.
+
+Note that as poking round chromium internal , it would default to { min: 0, max: 0 } , which would apply no restriction on udp port range.
+
+#### `contents.setWebRTCUDPPortRangePolicy(udpPortRange)`
```
Comment (jkleinsc) [inline docs/api/web-contents.md:2076]:
```suggestion
#### `contents.setWebRTCUDPPortRange(udpPortRange)`
```

File: docs/api/web-contents.md [hunk 10]
```diff
@@ -2064,6 +2064,23 @@ Setting the WebRTC IP handling policy allows you to control which IPs are
...
 more details.
 
+#### `contents.getWebRTCUDPPortRangePolicy()`
+
+Returns `Object`:
+
+* `min` Integer - The minimum port number of WebRTC UDP Port Range Policy.
+* `max` Integer - The maximum port number of WebRTC UDP Port Range Policy.
+
+Note that as poking round chromium internal , it would default to { min: 0, max: 0 } , which would apply no restriction on udp port range.
```
Comment (jkleinsc) [inline docs/api/web-contents.md:2074]:
```suggestion
By default this value is `{ min: 0, max: 0 }` , which would apply no restriction on the udp port range.
```

File: docs/api/web-contents.md [hunk 11]
```diff
@@ -2064,6 +2064,23 @@ Setting the WebRTC IP handling policy allows you to control which IPs are
 exposed via WebRTC. See [BrowserLeaks](https://browserleaks.com/webrtc) for
 more details.
 
+#### `contents.getWebRTCUDPPortRangePolicy()`
+
+Returns `Object`:
+
+* `min` Integer - The minimum port number of WebRTC UDP Port Range Policy.
+* `max` Integer - The maximum port number of WebRTC UDP Port Range Policy.
```
Comment (jkleinsc) [inline docs/api/web-contents.md:2072]:
```suggestion
* `min` Integer - The minimum UDP port number that WebRTC should use.
* `max` Integer - The maximum UDP port number that WebRTC should use.
```

File: docs/api/web-contents.md [hunk 12]
```diff
@@ -2064,6 +2064,23 @@ Setting the WebRTC IP handling policy allows you to control which IPs are
...
+
+Note that as poking round chromium internal , it would default to { min: 0, max: 0 } , which would apply no restriction on udp port range.
+
+#### `contents.setWebRTCUDPPortRangePolicy(udpPortRange)`
+
+* `udpPortRange` Object
+  * `min` Integer - The minimum port number of WebRTC UDP Port Range Policy.
+  * `max` Integer - The maximum port number of WebRTC UDP Port Range Policy.
```
Comment (jkleinsc) [inline docs/api/web-contents.md:2080]:
```suggestion
  * `min` Integer - The minimum UDP port number that WebRTC should use.
  * `max` Integer - The maximum UDP port number that WebRTC should use.
```

File: docs/api/web-contents.md [hunk 13]
```diff
@@ -2064,6 +2064,23 @@ Setting the WebRTC IP handling policy allows you to control which IPs are
...
+
+Setting the WebRTC UDP Port Range policy allows you to restrict the udp port range assigned to apps via WebRTC. Note that when setting to { min: 0, max: 0 }，no restriction is applied on udp port range.
```
Comment (jkleinsc) [inline docs/api/web-contents.md:2082]:
```suggestion
Setting the WebRTC UDP Port Range allows you to restrict the udp port range used by WebRTC.  By default the port range is unrestricted.  

*Note:** To reset to an unrestricted port range this value should be set to `{ min: 0, max: 0 }`.
```

File: spec/api-web-contents-spec.ts [hunk 14]
```diff
@@ -1288,6 +1288,53 @@ describe('webContents module', () => {
...
+      }).to.throw('Invalid arguments. Note that either port number should be in (0, 65535]');
+      expect(() => {
+        w.webContents.setWebRTCUDPPortRangePolicy({ min: 1, max: 65536 });
+      }).to.throw('Invalid arguments. Note that either port number should be in (0, 65535]');
+      expect(() => {
+        w.webContents.setWebRTCUDPPortRangePolicy({ min: 60000, max: 56789 });
+      }).to.throw('Invalid arguments. Note that min should not be greater than max');
+    });
+
+    it('setted webrtc udp port range policy can be set to default by input {min: 0, max: 0}', () => {
```
Comment (jkleinsc) [inline spec/api-web-contents-spec.ts:1323]:
```suggestion
    it('set webrtc udp port range policy can be reset to default by input {min: 0, max: 0}', () => {
```

**Review by jkleinsc — CHANGES_REQUESTED** [review]
Overall, looks good to me, but I think the method names should be shortened.  I don't think the word "Policy" is needed.  I've also added some suggestions for the documentation to make it clearer.

File: docs/api/web-contents.md (hunk 8 above)
Comment (wugaosheng123) [reply docs/api/web-contents.md:2067]: Thanks for your review:). 
I thought the udp port range config corresponds to the chrome://policy entry, and reference to another WebRtc api setWebRTCIPHandlingPolicy, I added the word "policy " in the API. 
But it is a bit long and tedious for the api. I will follow your suggestion to simplify it.

Comment (wanted002) [reply docs/api/web-contents.md:2067]: Thanks for your review:).
I thought the udp port range config corresponds to the chrome://policy entry, and reference to another WebRtc api setWebRTCIPHandlingPolicy, I added the word "policy " in the API.
But it is a bit long and tedious for the api. I will follow your suggestion to simplify it.

File: shell/browser/api/electron_api_web_contents.cc [hunk 15]
```diff
@@ -2539,6 +2539,60 @@ void WebContents::SetWebRTCIPHandlingPolicy(
...
+  auto* prefs = web_contents()->GetMutableRendererPrefs();
+
+  min = prefs->webrtc_udp_min_port;
+  max = prefs->webrtc_udp_max_port;
+
+  gin_helper::Dictionary dict = gin::Dictionary::CreateEmpty(isolate);
+  dict.Set("min", static_cast<uint32_t>(min));
+  dict.Set("max", static_cast<uint32_t>(max));
+  return dict.GetHandle();
+}
```
Comment (codebytere) [inline shell/browser/api/electron_api_web_contents.cc:2554]: We don't need to create a separate variable for this :)

```suggestion
v8::Local<v8::Value> WebContents::GetWebRTCUDPPortRange(
    v8::Isolate* isolate) const {
  auto* prefs = web_contents()->GetMutableRendererPrefs();

  gin_helper::Dictionary dict = gin::Dictionary::CreateEmpty(isolate);
  dict.Set("min", static_cast<uint32_t>(prefs->webrtc_udp_min_port));
  dict.Set("max", static_cast<uint32_t>(prefs->webrtc_udp_max_port));
  return dict.GetHandle();
}
```

File: shell/browser/api/electron_api_web_contents.cc [hunk 16]
```diff
@@ -2539,6 +2539,60 @@ void WebContents::SetWebRTCIPHandlingPolicy(
...
+            "Invalid arguments. Note that either port number should be in (0, "
+            "65535] or both are 0");
+    return;
+  }
+  if (min > max) {
+    gin_helper::ErrorThrower(args->isolate())
+        .ThrowError(
+            "Invalid arguments. Note that min should not be greater than max");
+    return;
+  }
```
Comment (codebytere) [inline shell/browser/api/electron_api_web_contents.cc:2581]: Let's throw slightly terser errors here, to match our typical errors a bit more.

```suggestion
  if (!args->GetNext(&range) || !range.Get("min", &min) ||
      !range.Get("max", &max)) {
    gin_helper::ErrorThrower(args->isolate())
        .ThrowError("'min' and 'max' are both required");
    return;
  }

  if ((0 == min && 0 != max) || max > UINT16_MAX) {
    gin_helper::ErrorThrower(args->isolate())
        .ThrowRangeError("'min' and 'max' must be in the (0, 65535] range or [0, 0]");
    return;
  }
  if (min > max) {
    gin_helper::ErrorThrower(args->isolate())
        .ThrowRangeError("'max' must be greater than or equal to 'min'");
    return;
  }
```

File: spec/api-web-contents-spec.ts [hunk 17]
```diff
@@ -1288,6 +1288,53 @@ describe('webContents module', () => {
...
 
+  describe('webrtc udp port range policy api', () => {
+    afterEach(closeAllWindows);
+    it('check default webrtc udp port range is {min: 0, max: 0}', () => {
+      const w = new BrowserWindow({ show: false });
+      const settings = w.webContents.getWebRTCUDPPortRange();
+      expect(settings).to.be.an('object');
+      expect(settings).to.have.property('min').that.is.equal(0);
+      expect(settings).to.have.property('max').that.is.equal(0);
+    });
```
Comment (codebytere) [inline spec/api-web-contents-spec.ts:1299]: We can simplify these assertions a bit:

```suggestion
    let w: BrowserWindow;

    beforeEach(() => {
      w = new BrowserWindow({ show: false });
    });

    afterEach(closeAllWindows);

    it('check default webrtc udp port range is { min: 0, max: 0 }', () => {
      const settings = w.webContents.getWebRTCUDPPortRange();
      expect(settings).to.deep.equal({ min: 0, max: 0 });
    });
```

File: spec/api-web-contents-spec.ts [hunk 18]
```diff
@@ -1288,6 +1288,53 @@ describe('webContents module', () => {
...
+
+    it('can set and get webrtc udp port range policy with correct arguments', () => {
+      const w = new BrowserWindow({ show: false });
+      w.webContents.setWebRTCUDPPortRange({ min: 1, max: 65535 });
+      const settings = w.webContents.getWebRTCUDPPortRange();
+      expect(settings).to.be.an('object');
+      expect(settings).to.have.property('min').that.is.equal(1);
+      expect(settings).to.have.property('max').that.is.equal(65535);
+    });
```
Comment (codebytere) [inline spec/api-web-contents-spec.ts:1308]:
```suggestion
     it('can set and get webrtc udp port range policy with correct arguments', () => {
      const w = new BrowserWindow({ show: false });
      w.webContents.setWebRTCUDPPortRange({ min: 1, max: 65535 });
      const settings = w.webContents.getWebRTCUDPPortRange();
      expect(settings).to.deep.equal({ min: 1, max: 65535 });
    });
```

File: spec/api-web-contents-spec.ts [hunk 19]
```diff
@@ -1288,6 +1288,53 @@ describe('webContents module', () => {
...
+      const settings = w.webContents.getWebRTCUDPPortRange();
+      expect(settings).to.be.an('object');
+      expect(settings).to.have.property('min').that.is.equal(1);
+      expect(settings).to.have.property('max').that.is.equal(65535);
+      w.webContents.setWebRTCUDPPortRange({ min: 0, max: 0 });
+      const defaultSetting = w.webContents.getWebRTCUDPPortRange();
+      expect(defaultSetting).to.be.an('object');
+      expect(defaultSetting).to.have.property('min').that.is.equal(0);
+      expect(defaultSetting).to.have.property('max').that.is.equal(0);
+    });
```
Comment (codebytere) [inline spec/api-web-contents-spec.ts:1335]:
```suggestion
    it('can reset webrtc udp port range policy to default with { min: 0, max: 0 }', () => {
      w.webContents.setWebRTCUDPPortRange({ min: 1, max: 65535 });
      const settings = w.webContents.getWebRTCUDPPortRange();
      expect(settings).to.deep.equal({ min: 1, max: 65535 });

      w.webContents.setWebRTCUDPPortRange({ min: 0, max: 0 });
      const defaultSetting = w.webContents.getWebRTCUDPPortRange();
      expect(defaultSetting).to.deep.equal({ min: 0, max: 0 });
    });
```

**Review by codebytere — COMMENTED** [review]
Some last cleanup comments and then I think we can get this merged! Thanks for working on this and being patient with us 😁

File: shell/browser/api/electron_api_web_contents.cc (hunk 16 above)
Comment (wanted002) [reply shell/browser/api/electron_api_web_contents.cc]: Sure~It seems better

File: spec/api-web-contents-spec.ts (hunk 17 above)
Comment (wanted002) [reply spec/api-web-contents-spec.ts]: Thanks for your guide, which is really helpful for a newbie like me.

**Review by codebytere — APPROVED** [review]
API LGTM!

**Review by jkleinsc — APPROVED** [review]
API LGTM

**Comment by siifan168** [comment]
May I ask, has this been released in the official version and can be used? How to use it

**Pull request merged**